### DISCUSSION TITLE
Is it possible to navigate away from a React Native screen with a query running in the background (aka who owns the query object)?

### DISCUSSION THREAD

**Discussion by tibbe** [TanStack/query#4276] [Q&A] [answered]
We're looking to hide some latency in an app by having a query execute asynchronously while we navigate to a different screen. It's a bit unclear to me who owns the object returned by `useQuery` and if it's safe to navigate before it finishes.
- If the screen still exists after navigating (e.g. because we push a new screen on the stack but the old one still exists), will the query continue running?
- If the screen no longer exists after navigating (e.g. because we pop it from the stack), will the query continue running?
- If the query continues running, will the callbacks for e.g. `onSuccess` still be invoked? (Of course we need to make sure that these callbacks don't refer to any state owned by the screen that no longer exists).

**Comment by TkDodo** [comment, marked as answer]
> If the screen still exists after navigating (e.g. because we push a new screen on the stack but the old one still exists), will the query continue running?

yes

> If the screen no longer exists after navigating (e.g. because we pop it from the stack), will the query continue running?


yes, unless you use [query cancellation](https://tanstack.com/query/v4/docs/guides/query-cancellation).

> If the query continues running, will the callbacks for e.g. onSuccess still be invoked? (Of course we need to make sure that these callbacks don't refer to any state owned by the screen that no longer exists).

callbacks are tied to the react lifecycle. If the component is mounted, they will be called, otherwise not. Mutations are different (but that wasn't the question)

**Comment by tibbe** [comment]
Thanks for the answer.

For mutations, will their callbacks (e.g. `onSuccess`) be called?

**Reply by TkDodo** [reply]
- the ones on `useMutation`: yes
- the ones on `mutate`: no
